bpf/lib/classifiers.h

```diff
@@ -0,0 +1,43 @@
+/* SPDX-License-Identifier: (GPL-2.0-only OR BSD-2-Clause) */
+/* Copyright Authors of Cilium */
+
+#pragma once
+
+#include "lib/common.h"
+
+/* Layer 3 packets are observed from the WireGuard device cilium_wg0 */
+#if defined(IS_BPF_WIREGUARD)
+# define CLASSIFIERS_DEVICE
+#endif
+
+typedef __u8 cls_flags_t;
+
+enum {
+	CLS_FLAG_IPV6	= (1 << 0),
+	CLS_FLAG_L3_DEV = (1 << 1),
+};
+
+#define EMPTY_CLASSIFIERS ((cls_flags_t)0)
+
+#ifdef CLASSIFIERS_DEVICE
+/* Compute packet layer classifiers
+ * - CLS_FLAG_L3_DEV: packet from a L3 device
+ * - CLS_FLAG_IPV6:   IPv6 packet when CLS_FLAG_L3_DEV is set. When already knowing a
+ *                    packet is IPv6, use ctx_device_classifiers6() instead.
+ */
+static __always_inline cls_flags_t
+ctx_device_classifiers(const struct __ctx_buff *ctx __maybe_unused)
+{
+	if (ctx->protocol == bpf_htons(ETH_P_IPV6))
+		return CLS_FLAG_L3_DEV | CLS_FLAG_IPV6;
+
+	return CLS_FLAG_L3_DEV;
```

Comment:
What happens if a developer accidentally uses `ctx_device_classifiers()` on an L2 netdev? Have you considered adding `if (ETH_HLEN == 0) { return CLS_FLAG_L3_DEV; }`?

Reply:
In theory, this returns `0` in case `IS_BPF_WIREGUARD` is not defined.
But yeah, could avoid the whole `#ifdef` logic and simply look at the `ETH_HLEN` 👍🏼